printing/proof_diffs.ml

```diff
@@ -317,31 +324,270 @@ let hyp_list_to_pp hyps =
   | h :: tl -> List.fold_left (fun x y -> x ++ cut () ++ y) h tl
   | [] -> mt ();;
 
-(* Special purpuse, use only for the IDE interface,  *)
-let diff_first_goal o_proof n_proof =
-  let first_goal_info proof =
-    match proof with
-    | None -> ([], StringMap.empty, Pp.mt ())
-    | Some proof2 ->
-      let (goals,_,_,_,sigma) = Proof.proof proof2 in
-      match goals with
-      | hd :: tl -> goal_info hd sigma;
-      | _ -> ([], StringMap.empty, Pp.mt ())
-  in
-  diff_goal_info (first_goal_info o_proof) (first_goal_info n_proof);;
-
-let diff_goals ?prev_gs n_gs =
-  let unwrap gs =
-    match gs with
-    | Some gs ->
-      let goal = Evd.sig_it gs in
-      let sigma = Refiner.project gs in
-      goal_info goal sigma
-    | None -> ([], StringMap.empty, Pp.mt ())
-  in
-  let (hyps_pp_list, concl_pp) = diff_goal_info (unwrap prev_gs) (unwrap n_gs) in
+let unwrap g_s =
+  match g_s with
+  | Some g_s ->
+    let goal = Evd.sig_it g_s in
+    let sigma = Refiner.project g_s in
+    goal_info goal sigma
+  | None -> ([], StringMap.empty, Pp.mt ())
+
+let diff_goal_ide og_s ng nsigma =
+  diff_goal_info (unwrap og_s) (goal_info ng nsigma)
+
+let diff_goal ?og_s ng ns =
+  let (hyps_pp_list, concl_pp) = diff_goal_info (unwrap og_s) (goal_info ng ns) in
   let open Pp in
   v 0 (
     (hyp_list_to_pp hyps_pp_list) ++ cut () ++
     str "============================" ++ cut () ++
     concl_pp);;
+
+
+(*** Code to determine which calls to compare between the old and new proofs ***)
+
+open Glob_term
+open Names
+open CAst
+
+let term_to_name x =
```

Comment:
I am not sure `glob_expr` should be used here; this PR certainly deserves a deeper look.

Reply:
This code was based on reverse engineering `show_proof`, the code for the `Show Proof` command, which is based entirely on same type that `x` has.  If that's good for `Show Proof`, why wouldn't it be good here?

BTW `x` is of type `'a glob_constr_r`; `glob_expr` is not (no longer?) a defined type in source tree.

Reply:
> If that's good for Show Proof, why wouldn't it be good here?

I'm much afraid the code for `Show Proof` may not be good.

Reply:
What data structure would you suggest using instead?

This code is far simpler than `Show Proof`--it's just a somewhat verbose tree traversal constructed mechanically from the type definition.  If `Show Proof` is using this data structure, it seems overzealous to say that diffs shouldn't.  If and when `Show Proof` is restructured, I can make the corresponding changes here.

The consequences of a failure to match or an incorrect match here are limited--the diff code will catch the failure to match, then print an error message and the proof step without diffs.  If there's an incorrect match, the diffs will look a bit odd, but the user can turn off diffs if that's distracting--and I'll make improvements.  `match_goals` is used only when multiple old goals have been rewritten, I suspect most proof steps don't do that.

Reply:
I didn't see where show proof uses a glob_constr , care to point that out?

Reply:
I stilll don't see where the `glob_constr` is used.

Reply:
> I stilll don't see where the `glob_constr` is used.

Sorry, I didn't see this question until just now.  In Constrextern.extern, the `c` in `match c with` is of type `glob_constr_r`.